### ISSUE TITLE
[UX] Redirect system-level messages (as opposed to form validation messages for example) to site admins and show friendly message to the user instead.

### ISSUE THREAD

**Issue by sutibun** [backdrop/backdrop-issues#1281]
This I am not 100% sure but I thought I read somewhere that you should always disable the display of error messages under "admin > config > development > logging"

![logging](https://cloud.githubusercontent.com/assets/14132746/10213908/71330f84-67d8-11e5-89ca-cd22d69ac0a3.png)

And seeing the note below says,

> It is recommended that sites running on production environments do not display any errors.

Should this be set to "none" by default? (and should I do the same?)

PS: Yay, finally moving one of my sites to Backdrop. Starting from scratch.

**Comment by serundeputy** [comment]
this seems like a reasonable [DX]/[UX] trade off to me.  If you are building a site (no code [`bad`] code) then this defaults to production ready.

If you are developing a site (`code` `code`); then you can turn it on and then off once you are ready to go live.

`PR` => https://github.com/backdrop/backdrop/pull/1102

**Comment by serundeputy** [comment]
@sutibun I just realized I ignored the question piece of your post.

I would say during development of the site leave this setting as `All messages` so you can catch problems during your development cycle(s).  Then once you are going to go live with the site set to `None`.

hope that helps,
~Geoff

**Comment by sutibun** [comment]
ah, sweet. Grazie @serundeputy

**Comment by quicksketch** [comment]
Thanks @sutibun, great question. Leaving messages on out-of-box is probably better for the health of our software (and in turn for the end-users). If all our software were perfect, this setting would make no difference, because it only controls application-level errors and warnings. That is, a validation warning or forgetting to fill out a required field may through an "error" message, but that isn't controlled by this option.

By leaving this option on, it makes it so that when a user does encounter a programming error, they'll get the message directly on the screen, rather than only logging it to the "Recent log entries", where they are likely to ignore or miss the error entirely. That makes it so that it's easier for them to report the error to the developers.

If you're running a production site, the advantage of turning it off is that you're not displaying environment information to your end-users, such as where you have installed Backdrop on the server. But it's a very minor concern. Disabling error messages is mostly for the sake of appearances

**Comment by sutibun** [comment]
> If you're running a production site, the advantage of turning it off is that you're not **displaying environment information to your end-users, such as where you have installed Backdrop on the server. But it's a very minor concern. _Disabling error messages is mostly for the sake of appearances_**

Is publicly showing where you installed Bd on the server bad? I just go by hearsay since I really don't know but I thought that was bad. In a Drupal post, I once pasted the exact server error message my site displayed (which included the directory I installed Drupal) and someone told me I might want to remove that info. I think he implied it might help hackers.

**Comment by klonos** [comment]
> I think he implied it might help hackers.

I'm pretty sure that that was precisely what he implied and it is true. It is good practice to remove the actual URL and path from the posts and start from the "root" ```/``` instead. For example ```/node/add``` instead of ```http://my.site.url/path/to/my/setup/directory/node/add```.

**Comment by sutibun** [comment]
> I'm pretty sure that that was precisely what he implied and it is true. It is good practice to remove the actual URL and path from the posts...

I don't know if that's the kind of error messages this option displays, but that was my thinking behind selecting "none" as a default option since new users wouldn't even look at this option and/or know it's potentially harmful. You guys would know much more than me on what's best as a default.

**Comment by klonos** [comment]
In Drupal, I used to use [Disable Messages](https://www.drupal.org/project/disable_messages), but the feature to [filter messages by module](https://www.drupal.org/node/1137722) I was always looking for never got implemented. It was blocked by [#1149006: Allow drupal_set_message to store information that can help define visibility (or not).](https://www.drupal.org/node/1149006) filed against core.

**Comment by sutibun** [comment]
> but the feature ... I was always looking for never got implemented. It was blocked by ...

hehe maybe it's me but seem you've gone through this kind of experience a lot

**Comment by klonos** [comment]
> hehe maybe it's me but seem you've gone through this kind of experience a lot

Yes, I was trying to hide "scary" failure messages from site visitors. There were even reports of certain users receiving errors about outdated core/modules on old, unmaintained websites. Very bad.

**Comment by klonos** [comment]
...actually...

> By leaving this option on, it makes it so that when a user does encounter a programming error, they'll get the message directly on the screen, rather than only logging it to the "Recent log entries", where they are likely to ignore or miss the error entirely. That makes it so that it's easier for them to report the error to the developers.

Honestly, no site visitor is ever going to register a GitHub account and report an error they received on a website - chances are they will avoid visiting the scary website again. Not good. The whole idea was to somehow "redirect" those messages to the site admin. If the site admin does not check their log file, then the errors would come up as something like this:

"Fatal error: your site just crashed on visitor ```[anonymous|username]``` with the following session ID ```[session_id]```, on ```[timestamp]```, while they were visiting ```[path]```, after submitting form ```[form_id]```"

The site admin/developer is more likely to investigate the error and then report it to us.

**Comment by sutibun** [comment]
> "Fatal error: your site just crashed on visitor [anonymous|username] with the following session ID [session_id], on [timestamp], while they were visiting [path], after submitting form [form_id]"
> 
> The site admin/developer is more likely to investigate the error and then report it to us.

Since some site admins/dev don't constantly check the "recent log entries", how about highlighting urgent messages like these on the dashboard #495 so it'll catch their attention. (preferably on the #639 What? I'm a promoter)

**Comment by klonos** [comment]
...all the simple user (site visitor) should get is a friendly message like this:

"Oops, something went wrong there. It was not your fault and the error has been reported to our support team. Sorry about that. For your reference, the issue number is [some_error_log_id]"

**Comment by klonos** [comment]
...was about to mention #639 but the Monkey beat me to it again :wink:

**Comment by klonos** [comment]
Better? :wink:

**Comment by sutibun** [comment]
Would say so...

-----
**Nate:** sigh. I won't hear the end of #639 until it's built. Fine. _\*begins typing code\*_
**monkey and Klon:** ^_______^

**Comment by klonos** [comment]
Yeah, unfortunately, seems like the Monkey and myself are in the same spot where we simply cannot put code where our mouth is. We only have ideas to share.

**Comment by sutibun** [comment]
speak for yourself. I can also provide bananas... and cheering.

**Comment by klonos** [comment]
:smile:

Some monkeys throw poop too when upset. I hope you are not one of those.

**Comment by sutibun** [comment]
er... I'd never (^\_^;

**Comment by quicksketch** [comment]
:laughing: :heart:

Thanks guys. There's a lot going on here, are there any direct suggestions that we should follow through on? We've already got the notification issue at #639. If that were implemented, we would still need a way to get a message into the status report saying there was a problem. We could potentially log the count of the highest level errors (WATCHDOG_CRITICAL and WATCHDOG_ERROR) and then add a status report message if a threshold of errors is reached (say like 100 a day). Then we'd be directing users to the watchdog log from the status report.

**Comment by klonos** [comment]
#639 is for getting a badge shown on the admin bar (available when logged in as a site admin). This issue here is to implement a generic "Oops, something went wrong. The site admin has been notified." error shown to site visitors while the actual, "scary" ```PDOException```/```Fatal error``` messages are logged and the admin notified (via #639).

**Comment by klonos** [comment]
...the "opposed" part of the issue title is to clarify that this mechanism should only kick in for system errors - validation errors (valid email form in an email field for example) will still be shown to the end user.

**Comment by sutibun** [comment]
what he said
